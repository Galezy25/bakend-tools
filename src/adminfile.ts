import fs from 'fs';
import path from 'path';

export class AdminFile {
    public static get path() {
        return path;
    }

    protected _rootPath: string;

    public get rootPath() {
        return this._rootPath;
    }

    /**
     *
     * @param rootPath Absolute path of the root of this file manager.
     */
    constructor(rootPath: string) {
        this._rootPath = rootPath;
        if (!fs.existsSync(this._rootPath)) {
            fs.mkdirSync(this._rootPath);
        }
    }

    /**
     *
     * @param dir Directory where will be write the file
     * @param namedWithExt Name with extension of the file
     * @param data Content of the file
     * @returns Promise that returns relativePath
     */
    public async writeFile(dir: string, namedWithExt: string, data: any) {
        let fullDir = path.resolve(this._rootPath, dir);
        this.makeDir(fullDir);
        let fullPath = path.resolve(fullDir, namedWithExt);
        let relativePath = path.relative(this._rootPath, fullPath);
        fs.writeFileSync(fullPath, data);
        return relativePath;
    }

    /**
     *
     * @param dir Directory where is the file to append data
     * @param namedWithExt Name with extension of the file
     * @param data Content to append in the file
     * @returns Promise that returns relativePath
     */
    async appendFile(dir: string, namedWithExt: string, data: any) {
        let fullDir = path.resolve(this._rootPath, dir);
        this.makeDir(fullDir);
        let fullPath = path.resolve(fullDir, namedWithExt);
        let relativePath = path.relative(this._rootPath, fullPath);
        fs.appendFileSync(fullPath, data);
        return relativePath;
    }

    /**
     *
     * @param dir Directory where find the file to read.
     * @param namedWithExt Name with extension of the file.
     * @returns Promise that resolve with a string of the content of the file.
     */
    public async readFile(dir: string, namedWithExt: string) {
        let afPath = this._rootPath;
        let pathFile = path.resolve(afPath, dir, namedWithExt);
        let data = fs.readFileSync(pathFile, 'utf8');
        return data;
    }

    /**
     *
     * @param dir Directory to search into
     * @param named Segment or name of the file without extension.
     * @returns Promise that resolves with the full path and the name with extension of the file.
     */
    public async searchFile(dir: string, named: string) {
        let afPath = this._rootPath;
        let old_path = path.resolve(afPath, dir);
        let files = fs.readdirSync(old_path);
        let namedWithExt = files
            .filter(filename => filename.includes(named))
            .sort((a, b) => a.length - b.length)[0];
        if (!namedWithExt) {
            const error = new Error('File not found or does not exist');
            error.name = 'NOT_FOUND';
            throw error;
        }
        let absolutePath = path.resolve(afPath, dir, namedWithExt);
        return {
            absolutePath,
            namedWithExt,
        };
    }

    /**
     *
     * @param dir Directory of the file.
     * @param namedWithExt Name with extension of the file.
     * @returns Promise
     */
    public async eraseFile(dir: string, namedWithExt: string) {
        let afPath = this._rootPath;

        let old_path = path.join(afPath, dir, namedWithExt);
        fs.unlinkSync(old_path);
        return;
    }

    /**
     *
     * @param dir Directory to search into.
     * @param named Segment or name of the file without extension.
     */
    public async searchEraseFile(dir: string, named: string) {
        try {
            const { namedWithExt } = await this.searchFile(dir, named);
            await this.eraseFile(dir, namedWithExt);
            return;
        } catch (err) {
            if ((err as any)?.name !== 'NOT_FOUND') throw err;
        }
    }

    /**
     *
     * @param dir Create directory if not exist.
     */
    public makeDir(dir: string) {
        if (!fs.existsSync(path.resolve(this._rootPath, dir))) {
            fs.mkdirSync(path.resolve(this._rootPath, dir));
        }
    }
}

export default AdminFile;
