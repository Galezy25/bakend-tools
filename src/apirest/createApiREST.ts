import express from 'express';

type RESTfulHandler = express.RequestHandler | express.RequestHandler[];

/**
 *
 * @param handlers
 * @param idName
 * @returns
 */
export function createApiREST(
    handlers: {
        /**
         * GET: `/`
         */
        find?: RESTfulHandler;

        /**
         * GET: `/:idName`
         */
        findOne?: RESTfulHandler;

        /**
         * GET: `/count`
         */
        count?: RESTfulHandler;

        /**
         * POST: `/`
         */
        create?: RESTfulHandler;

        /**
         * PUT: `/:idName`
         */
        replace?: RESTfulHandler;

        /**
         * PATCH: `/:idName`
         */
        modify?: RESTfulHandler;

        /**
         * DELETE: `/:idName`
         */
        remove?: RESTfulHandler;
    },
    idName?: string
): express.Router {
    const router = express.Router();
    if (handlers.find) {
        router.get('/', ...getHandlers(handlers.find));
    }
    if (handlers.count) {
        router.get(`/count`, ...getHandlers(handlers.count));
    }
    if (handlers.findOne) {
        if (!idName) throw new Error('idName needed');
        router.get(`/:${idName}`, ...getHandlers(handlers.findOne));
    }
    if (handlers.create) {
        router.post('/', ...getHandlers(handlers.create));
    }
    if (handlers.replace) {
        if (!idName) throw new Error('idName needed');
        router.put(`/:${idName}`, ...getHandlers(handlers.replace));
    }
    if (handlers.modify) {
        if (!idName) throw new Error('idName needed');
        router.patch(`/:${idName}`, ...getHandlers(handlers.modify));
    }
    if (handlers.remove) {
        if (!idName) throw new Error('idName needed');
        router.delete(`/:${idName}`, ...getHandlers(handlers.remove));
    }
    return router;
}

const getHandlers = (handlers: RESTfulHandler) =>
    handlers instanceof Array ? handlers : [handlers];
