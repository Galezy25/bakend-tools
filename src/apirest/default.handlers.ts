import express from 'express';

import CRUD from '../crud.interface';

export const defaultFind = (collection: CRUD): express.RequestHandler => async (
    req,
    res,
    next
) => {
    try {
        let data = await collection.find({ ...req.query, ...req.params });
        res.json(data);
    } catch (error) {
        next(error);
    }
};

export const defaultFindOne = (
    collection: CRUD
): express.RequestHandler => async (req, res, next) => {
    try {
        let data = await collection.findOne({ ...req.query, ...req.params });
        res.json(data);
    } catch (error) {
        next(error);
    }
};

export const defaultCount = (
    collection: CRUD
): express.RequestHandler => async (req, res, next) => {
    try {
        let data = await collection.count({ ...req.query, ...req.params });
        res.json(data);
    } catch (error) {
        next(error);
    }
};

export const defaultCreate = (
    collection: CRUD,
    idName?: string
): express.RequestHandler => async (req, res, next) => {
    try {
        let data: any = await collection.create(req.body);
        const { insertId } = data;
        if (idName) {
            data = await collection.findOne({ [idName]: insertId });
        }
        res.status(201).json(data);
    } catch (error) {
        next(error);
    }
};

export const defaultModify = (
    collection: CRUD
): express.RequestHandler => async (req, res, next) => {
    try {
        await collection.update({ ...req.query, ...req.params }, req.body);
        res.sendStatus(200);
    } catch (error) {
        next(error);
    }
};

export const defaultReplace = (
    collection: CRUD
): express.RequestHandler => async (req, res, next) => {
    try {
        await collection.replace({ ...req.body, ...req.params });
        res.sendStatus(200);
    } catch (error) {
        next(error);
    }
};

export const defaultRemove = (
    collection: CRUD
): express.RequestHandler => async (req, res, next) => {
    try {
        await collection.delete({ ...req.query, ...req.params });
        res.sendStatus(200);
    } catch (error) {
        next(error);
    }
};
