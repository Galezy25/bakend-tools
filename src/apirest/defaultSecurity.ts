import { IncomingHttpHeaders } from 'http';
import express from 'express';
import { ErrorHTTP } from '../error.handler';

export const defaultSecurity = (
    verifyToken: (token: string) => Promise<any>
) => (
    validateData: (payloadToMatch: Matcher) => boolean = () => true
) => async (req: express.Request, _res: any, next: express.NextFunction) => {
    try {
        if (!req.headers.authorization) {
            return next(
                new ErrorHTTP('No token authorization', {
                    statusCode: 401,
                })
            );
        }

        const payload = await verifyToken(req.headers.authorization);
        const payloadToMatch: Matcher = {
            ...payload,
            _params: req.params,
            _query: req.query,
            _headers: req.headers,
            _cookies: req.cookies,
        };

        if (!validateData(payloadToMatch)) {
            next(new ErrorHTTP('No match', { statusCode: 403 }));
        }
        next();
    } catch (error) {
        next(error);
    }
};

export interface Matcher {
    /**
     * Property into payload of the token.
     */
    [property: string]: any;
    /**
     * Request params.
     */
    _params?: { id: string };
    /**
     * Request query
     */
    _query?: any;
    /**
     * Request headers
     */
    _headers?: IncomingHttpHeaders;
    /**
     * Request cookies
     */
    _cookies?: any;
}
