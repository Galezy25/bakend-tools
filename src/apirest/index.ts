export * from './createApiREST';
export * from './default.handlers';
export * from './defaultSecurity';
