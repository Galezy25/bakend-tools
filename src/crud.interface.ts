export interface CRUD {
    id_name: string;
    find: <T = any[]>(context: { [key: string]: any }) => Promise<T>;
    findOne: <T = any>(context: { [key: string]: any }) => Promise<T>;
    count: (context: { [key: string]: any }) => Promise<number>;
    create: (
        valuesToCreate: any
    ) => Promise<{
        [key: string]: any;
        insertId: any;
    }>;
    update: (
        context: { [key: string]: any },
        valuesToUpdate: any
    ) => Promise<any>;
    replace: (valuesToReplace: any) => Promise<any>;
    delete: (context: { [key: string]: any }) => Promise<any>;
}

export default CRUD;
