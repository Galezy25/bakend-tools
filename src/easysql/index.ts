import * as Constraint from './constraint';
export * from './connection';
export * from './table';
export * from './column';
export { Constraint };
