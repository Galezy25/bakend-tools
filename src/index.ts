import * as EasySQL from './easysql';
import * as ApiREST from './apirest';
export * from './adminfile';
export * from './simpletoken';
export * from './error.handler';
export * from './file.api';
export * from './restful.api';
export * from './crud.interface';
export * from './permissions.manager';
export * from './session.manager';
export { EasySQL, ApiREST };
