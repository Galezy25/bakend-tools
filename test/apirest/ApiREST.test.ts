import express from 'express';
import { NextFunction } from 'express-serve-static-core';
import request = require('supertest');
import Connection from '../../src/easysql/connection';
import * as ApiREST from '../../src/apirest';
import SimpleToken from '../../src/simpletoken';
import CRUD from '../../src/crud.interface';

const mockQuery = jest.fn(
    (
        query: string,
        callback: (err: any, results: any, fields: any) => void
    ) => {
        if (query.includes('INSERT INTO')) {
            callback(
                undefined,
                {
                    insertId: 265,
                },
                []
            );
        } else {
            callback(
                undefined,
                [
                    {
                        id: 265,
                        name: 'Atari 2600',
                        cost: 849.88,
                    },
                ],
                []
            );
        }
    }
);
jest.mock('mysql', () => {
    return {
        createPool: jest.fn(_config => ({
            query: mockQuery,
            end: jest.fn(),
        })),
        escape: jest.fn((value: any) => {
            return typeof value === 'string' ? `'${value}'` : value + '';
        }),
    };
});

describe('ApiREST tests', () => {
    const body = {
        id: 265,
        name: 'Atari 2600',
        cost: 849.88,
    };
    const confirmationMd5 = 'Confirmation pass';
    const connection = new Connection({
        user: 'tester',
        database: 'test',
    });
    const table = connection.table('products', {
        id_name: 'id',
        tablesRelated: ['categories'],
    });
    const errorCRUD: CRUD = {
        id_name: 'id',
        count() {
            throw new Error('Error');
        },
        find() {
            throw new Error('Error');
        },
        findOne() {
            throw new Error('Error');
        },
        create() {
            throw new Error('Error');
        },
        update() {
            throw new Error('Error');
        },
        replace() {
            throw new Error('Error');
        },
        delete() {
            throw new Error('Error');
        },
    };
    const simpleToken = new SimpleToken('RESTful tests');
    const getSecurityMiddleware = ApiREST.defaultSecurity(simpleToken.verify);
    const defaultRESTful = ApiREST.createApiREST(
        {
            find: ApiREST.defaultFind(table),
            findOne: ApiREST.defaultFindOne(table),
            count: ApiREST.defaultCount(table),
            create: [
                getSecurityMiddleware(payload => {
                    return payload.modify;
                }),
                ApiREST.defaultCreate(table, 'id'),
            ],
            modify: [
                getSecurityMiddleware(payload => {
                    return payload.modify;
                }),
                ApiREST.defaultModify(table),
            ],
            replace: [
                getSecurityMiddleware(payload => {
                    return payload.modify;
                }),
                ApiREST.defaultReplace(table),
            ],
            remove: [
                getSecurityMiddleware(payload => {
                    return (
                        payload.modify &&
                        !!payload._headers &&
                        payload._headers['x-confirmation'] === confirmationMd5
                    );
                }),
                ApiREST.defaultRemove(table),
            ],
        },
        'id'
    );
    const onlyFindRESTful = ApiREST.createApiREST(
        {
            find: ApiREST.defaultFind(table),
        },
        'id'
    );
    const errorRESTful = ApiREST.createApiREST(
        {
            find: ApiREST.defaultFind(errorCRUD),
            findOne: ApiREST.defaultFindOne(errorCRUD),
            count: ApiREST.defaultCount(errorCRUD),
            create: ApiREST.defaultCreate(errorCRUD),
            replace: ApiREST.defaultReplace(errorCRUD),
            modify: ApiREST.defaultModify(errorCRUD),
            remove: ApiREST.defaultRemove(errorCRUD),
        },
        'id'
    );
    const tokenCanModify = simpleToken.sign({ modify: true });
    const tokenCantModify = simpleToken.sign({});
    const app = express();
    app.use(express.json());
    app.use('/default/', defaultRESTful);
    app.use('/resources/onlyFind/', onlyFindRESTful);
    app.use('/resources/errors/', errorRESTful);

    app.use(
        (err: any, __req: any, res: express.Response, next: NextFunction) => {
            if (err.statusCode <= 500) {
                res.sendStatus(err.statusCode);
            } else {
                next(err);
            }
        }
    );

    beforeEach(() => {
        mockQuery.mockClear();
    });

    test('createApiREST error to create "idName needed"', async () => {
        expect(() =>
            ApiREST.createApiREST({
                findOne: () => {},
            })
        ).toThrow();
        expect(() =>
            ApiREST.createApiREST({
                modify: () => {},
            })
        ).toThrow();
        expect(() =>
            ApiREST.createApiREST({
                replace: () => {},
            })
        ).toThrow();
        expect(() =>
            ApiREST.createApiREST({
                remove: () => {},
            })
        ).toThrow();
    });

    test('Default find, findOne, count tests', async () => {
        await request(app)
            .get('/default?_sort=cost:ASC&_limit=50&_start=250')
            .expect(200);
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();

        await request(app)
            .get(
                '/default/265?categories_inner=this.category:categories.id,name:category_name'
            )
            .expect(200);
        expect(mockQuery.mock.calls[1][0]).toMatchSnapshot();

        await request(app)
            .get('/default/count?category_in=1,2,3,4')
            .expect(200);
        expect(mockQuery.mock.calls[2][0]).toMatchSnapshot();
    });

    test('Default create and payload security tests', async () => {
        await request(app)
            .post('/default')
            .send(body)
            .set('authorization', tokenCantModify)
            .expect(403);
        await request(app)
            .post('/default')
            .send(body)
            .set('authorization', tokenCanModify)
            .expect(201);
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();
    });

    test('Default modify tests', async () => {
        await request(app)
            .patch('/default/265')
            .set('authorization', tokenCanModify)
            .send({ cost: 850 })
            .expect(200);
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();
    });

    test('Default replace tests', async () => {
        await request(app)
            .put('/default/265')
            .set('authorization', tokenCanModify)
            .send(body)
            .expect(200);
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();
    });

    test('Default remove and headers tests', async () => {
        await request(app)
            .delete('/default/265')
            .set('authorization', tokenCanModify)
            .expect(403);
        await request(app)
            .delete('/default/265')
            .set('authorization', tokenCanModify)
            .set('x-confirmation', confirmationMd5)
            .expect(200);
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();
    });

    test('OnlyFind test', async () => {
        await request(app)
            .get('/resources/onlyFind')
            .expect(200);
        await request(app)
            .get('/resources/onlyFind/265')
            .expect(404);
        await request(app)
            .post('/resources/onlyFind')
            .expect(404);
        await request(app)
            .patch('/resources/onlyFind/265')
            .expect(404);
        await request(app)
            .delete('/resources/onlyFind/265')
            .expect(404);
    });

    test('Errors test', async () => {
        await request(app)
            .get('/resources/errors')
            .expect(500);
        await request(app)
            .get('/resources/errors/265')
            .expect(500);
        await request(app)
            .post('/resources/errors')
            .expect(500);
        await request(app)
            .put('/resources/errors/265')
            .expect(500);
        await request(app)
            .patch('/resources/errors/265')
            .expect(500);
        await request(app)
            .delete('/resources/errors/265')
            .expect(500);
        await request(app)
            .get('/resources/errors/count')
            .expect(500);
    });
});
