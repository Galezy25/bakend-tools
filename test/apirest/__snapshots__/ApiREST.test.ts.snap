// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiREST tests Default create and payload security tests 1`] = `"INSERT INTO products (id,name,cost) VALUES (265, 'Atari 2600', 849.88);"`;

exports[`ApiREST tests Default find, findOne, count tests 1`] = `"SELECT this.* FROM products this ORDER BY cost ASC LIMIT 250, 50;"`;

exports[`ApiREST tests Default find, findOne, count tests 2`] = `"SELECT this.*, categories.name AS category_name FROM products this INNER JOIN categories ON this.category = categories.id WHERE id = '265';"`;

exports[`ApiREST tests Default find, findOne, count tests 3`] = `"SELECT COUNT(*) as count FROM products this WHERE category IN ('1','2','3','4')"`;

exports[`ApiREST tests Default modify tests 1`] = `"UPDATE products SET cost=850  WHERE id = '265'"`;

exports[`ApiREST tests Default remove and headers tests 1`] = `"DELETE FROM products  WHERE id = '265'"`;

exports[`ApiREST tests Default replace tests 1`] = `"REPLACE INTO products (id,name,cost) VALUES ('265', 'Atari 2600', 849.88);"`;
